backends/arm/passes/permute_memory_pass.py

```diff
@@ -0,0 +1,55 @@
+# Copyright 2024 Arm Limited and/or its affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import torch
+from executorch.backends.arm.passes.arm_pass import ArmPass
+from executorch.backends.arm.tosa_quant_utils import q_op
+from executorch.backends.arm.tosa_utils import (
+    is_bias_node_for_quantized_addmm,
+    is_consumer_node_depthwise_conv2d,
+)
+from executorch.exir.dialects._ops import ops as exir_ops
+from executorch.exir.pass_base import PassResult
+
+
+class PermuteMemoryPass(ArmPass):
+
+    def is_weight_node_for_dw_conv(self, node: torch.fx.Node):
+        if node.name in self.exported_program.graph_signature.inputs_to_parameters:
+            consumer_node = list(node.users)[0]
+            if not is_bias_node_for_quantized_addmm(node) and (
+                not consumer_node.target == exir_ops.edge.aten.convolution.default
+                or not list(consumer_node.users)[0].target == q_op
+            ):
+                return is_consumer_node_depthwise_conv2d(node)
+        elif node.name in self.exported_program.graph_signature.inputs_to_buffers:
+            return is_consumer_node_depthwise_conv2d(list(node.users)[0])
+        elif node.op == "call_function" and is_consumer_node_depthwise_conv2d(node):
+            if len(node.args) > 0:
+                return (
+                    node.args[0].name
+                    in self.exported_program.graph_signature.inputs_to_buffers
+                )
+        return False
+
+    def call(self, graph_module: torch.fx.GraphModule):
+        NHWC_Order = (0, 2, 3, 1)
+        HWCM_Order = (2, 3, 0, 1)
+        for node in graph_module.graph.nodes:
+            if type(node.meta["val"]) is tuple:
+                data = node.meta["val"][0].data
+            else:
+                data = node.meta["val"].data
+
+            if len(data.shape) == 4:
+                dim_order = NHWC_Order
+                if self.is_weight_node_for_dw_conv(node):
+                    dim_order = HWCM_Order
+            else:
+                dim_order = tuple(range(len(data.shape)))
+            node.meta["tosa_dim_order"] = dim_order
```

Comment:
Let me think, there are at least three ways of doing this. 

(1) the way you are doing it. i.e. in `preprocess()` insert a dim_order tuple, track it manually and transform things as we convert to TOSA.
:+1: Flexibility
:+1: Mostly transparent to the user, expect the compile_spec
:-1: You are expecting the input to the `call_delegate()` to be in NHWC format
:-1: Doesn't use pytorch infra 
:-1: `node.meta['tosa_dim_order'] != node.meta["val"].dim_order()` which may cause confusion


(2) Global switch, i.e. `gm = export(model.to(memory_format=torch.channels_last), torch.randn(1,2,3,4).to(memory_format=torch.channels_last)`. This way `gm` should have everything in nhwc dim_order.
:+1: Clean impl, uses pytorch infra
:-1: If something falls out of delegate, that must be run in NHWC format
:-1: Assumes input conversion to NHWC happens outside of the model. Might be OK if you are getting it already in NHWC.

(3) Runtime conversion: After export we insert the `to_dim_order(NHWC)` ops in the beginning/end (also on the weights/placeholders etc.) as we want. Consume these nodes as we see fit in the delegate.
:+1: Mix of (1) and (2), allows you to select where the input conversion to NHWC happens.
:+1: Makes it flexible in terms of which node runs in which format. That is if you leave the node, it will take NCHW and convert it at runtime on the CPU. If you partition and delete the node the input to the delegate and to the model (I think, I have to check, this is similar to deleting first q node) will be NHWC.
:+1: Still uses pytorch infra, i.e. no need to manually maintain `tosa_dim_order` unless for cases like DW conv (but I see that as a special constraint from the backend so..)
:-1: slightly harder to implement, particularly if you are sprinkling multiple nodes in the graph. But if you run this pass before partitioner then it might not be too difficult.
:-1: Potentially involves running a pass outside i.e. modifying the graph before `to_backend()`. As long as this is user visible it might be ok.
:-1: Exposes some footguns when multiple delegates are involved but should be the case here.

Reply:
Thank you Digant for your comments, much appreciated!

(1) This would assume that the input is in NHWC-format if `permute_memory_to_nhwc==True`, yes. The next step of this PR could be to add a pass or extend this pass to add to_dim_order-ops around all our delegates such that the graph becomes: 
Input (default format) -> ... -> to_dim_order(nhwc) -> delegate -> to_dim_order(default format) -> ... -> output (default format). This would require that we modify the graph outside (just before and after) our delegate after partitioning. What's the proper way of doing this?
I agree with the last two bullets. Those are the drawbacks of the current solution.

(2) `model.to(memory_format=torch.channels_last)` and `node.meta["val"].to(memory_format=torch.channels_last)` gives `dim_order=(0,1,2,3)` for certain shapes such as (3,1,3,3). Otherwise I think `node.meta["val"].to(memory_format=torch.channels_last)` could have been used in (1) instead of annotating `node.meta["tosa_dim_order"]`.

(3) I don't think I fully understand this solution. We insert to_dim_order-ops at the inputs, outputs and placeholders of the full graph or just our delegate sub-graph? Will the dim_order be propagated forward when inserting to_dim_order-ops?